### DISCUSSION TITLE
Wrong env variable in the production middleware

### DISCUSSION THREAD

**Discussion by youminkim** [vercel/next.js#49575] [Help] [answered]
### Summary

We use two Upstash redis instances for 1) prod and 2) preview. We use Vercel Environment Variables `UPSTASH_EMO_REST_URL` and have set differently. 

In our production deployments, usual serverless functions shows correct env variable. However, in middleware (edge), it shows (2) preview's variable. 

I am not sure if this is intended

**Comment by youminkim** [comment]
+We use Singapore region (sin1) serverless function. 

Deployments 
> Running build in San Francisco, USA (West) – sfo1

**Comment by youminkim** [comment]
After few trials, this is what I found

1) The edge's env variable does not change even when I changed the preview env variable in Vercel dashboard
2) When I created a new env variable in Vercel dashboard, it showed same issue again.

I ended up setting same value in prod / preview.

**Reply by smaeda-ks** [reply]
@youminkim can you paste your project id here? Also, any chance you're using monorepo tools, such as Turborepo or Nx? Those tools require your environment variables to be defined in the config file accordingly to calc cache hash properly:

- https://vercel.com/docs/concepts/monorepos/turborepo#handling-environment-variables
- https://nx.dev/concepts/how-caching-works#runtime-hash-inputs

**Reply by mmachatschek** [reply]
@smaeda-ks we are using turborepo

~even rebuilding the deployment without build cache doesn't solve the problem.~

~moving the code from the middleware into a handler does work. The edge function is the problem on our side.~

**Reply by mmachatschek** [reply]
@smaeda-ks adding environment variables to the turbo config would invalidate our build caches and deploying to production would then invalidate the build cache. I guess this is not feasible for us as would not get builds from the turbo repo cache

are environment variables inlined into the edge functions?

**Reply by smaeda-ks** [reply]
it should be safe to add non-public env vars to your config (see the note in the section below):
https://turbo.build/repo/docs/core-concepts/caching#automatic-environment-variable-inclusion

**Reply by youminkim** [reply]
Hi @smaeda-ks project id is `emo`. We do not use monorepo

**Reply by javivelasco** [reply]
@youminkim or @mmachatschek can you share a reproduction or the deployment where you are getting an unexpected value? Feel free to reach out through Twitter DM for this so we can investigate further please.

**Reply by ccjvp** [reply]
@javivelasco We're also experiencing the same same issue with our project (ID: prj_lqnu9ls9hC710KOR24f9WFXt8pW8) in e.g. the latest main branch deployment (ID: dpl_FUJJ9xs3jdjChaqMm5c5wAmGCLMd).

**Reply by javivelasco** [reply]
@ccjvp That project seems to have no ENV vars defined. Can you provide a URL where you access the ENV that has the unexpected value?

**Reply by ccjvp** [reply]
@javivelasco That's weird, I can see env vars on the Vercel dashboard for that project. Here's a URL https://softserve-fv3jquxr8-mjuk.vercel.app where the middleware accesses ENV vars with unexpected (preview) values.

**Reply by javivelasco** [reply, marked as answer]
Identified the issue. We are working on a fix

**Reply by javivelasco** [reply]
It is fixed now

**Comment by mmachatschek** [comment]
We are facing the same issue

**Comment by javivelasco** [comment]
@mmachatschek @ccjvp @youminkim I apologize for the delay. We've just rolled a fix and your affected deployments should be working fine. Can you please verify that this is the case?

**Reply by youminkim** [reply]
Hi @javivelasco, I just redeployed a recent preview deployment after changing an env variable for preview. However, the edge's env variable is not change (wrong). I confirmed that serverless function's env variable is properly changed. 

Project id: emo
Deoloyment id: FN2TDQ13fZYBbpod6ATydoLgddbx

**Reply by smaeda-ks** [reply]
Do you have any `.env` files committed in your repository?

**Reply by ccjvp** [reply]
@javivelasco No problem, thanks for looking into it! Our deployments now have the right env vars 👍

**Reply by youminkim** [reply]
@smaeda-ks No we did not

**Reply by youminkim** [reply]
I checked my most recent deployments today but still edge does not reflect right env variable yet.

**Reply by smaeda-ks** [reply]
@youminkim In your project `emo`, I don't see `UPSTASH_EMO_REST_URL` is set to the project's environment variables. Where are you defining it?

**Reply by youminkim** [reply]
@smaeda-ks I just checked again and I confirmed that it is properly set. Thanks for the fix!
